tests/php/Functional/wc/payment/OrderItemsRefunderTest.php

```diff
@@ -0,0 +1,436 @@
+         * Expectations
+         */
+        $remoteOrder->expects($this->never())->method('cancelLines');
+        $remoteOrder->expects($this->never())->method('refund');
+
+        expectedActionDone(OrderItemsRefunder::ACTION_AFTER_CANCELED_ORDER_ITEMS)->never();
+        expectedActionDone(OrderItemsRefunder::ACTION_AFTER_REFUND_ORDER_ITEMS)->never();
+
+        /*
+         * Execute Test
+         */
+        $orderItemsRefunder->refund($order, [$orderItem], $remoteOrder, $refundReason);
+    }
+
+    /**
+     * @throws ApiException
+     * @throws UnexpectedValueException
+     * @throws PHPUnit_Framework_Exception
+     */
+    public function testUnexpectedValueExceptionWhenBuildRefundItems()
+    {
+        /*
+         * Stubs
+         */
+        $order = new WC_Order();
+        // Passing null is the key here, this will throw the exception because of invalid value.
+        /** @var WC_Order_Item $orderItem */
+        $orderItem = $this->orderItem(['meta' => null]);
+        $orderLineItem = $this->orderLineItem(
+            [
+                'metadata' => ['order_item_id' => $orderItem->get_meta()],
+                'id' => uniqid(),
+            ]
+        );
+        $remoteOrder = $this->remoteOrder([$orderLineItem]);
+        $refundReason = uniqid();
+
+        /*
+         * Sut
+         */
+        $orderItemsRefunder = new OrderItemsRefunder(
+            $this->refundLineItemsBuilder,
+            $this->dataHelper,
+            $this->ordersApiClient
+        );
+
+        /*
+         * Expectations
+         */
+        $this->expectException(UnexpectedValueException::class);
+        $this->expectExceptionMessage(
+            'One of the WooCommerce order items does not have the refund item id meta value associated to Mollie Order item.'
+        );
+
+        /*
+         * Execute Test
+         */
+        $orderItemsRefunder->refund($order, [$orderItem], $remoteOrder, $refundReason);
+    }
+
+    /**
+     * @throws ApiException
+     * @throws PHPUnit_Framework_Exception
+     * @throws UnexpectedValueException
+     */
+    public function testUnexpectedValueExceptionWhenBuildRemoteItems()
+    {
+        /*
+         * Stubs
+         */
+        $order = new WC_Order();
+        /** @var WC_Order_Item $orderItem */
+        $orderItem = $this->orderItem(['meta' => uniqid()]);
+        $orderLineItem = $this->orderLineItem(
+            [
+                'metadata' => null,
+                'id' => uniqid(),
+            ]
+        );
+        $remoteOrder = $this->remoteOrder([$orderLineItem]);
+        $refundReason = uniqid();
+
+        /*
+         * Sut
+         */
+        $orderItemsRefunder = new OrderItemsRefunder(
+            $this->refundLineItemsBuilder,
+            $this->dataHelper,
+            $this->ordersApiClient
+        );
+
+        /*
+        * Expectations
+        */
+        $this->expectException(UnexpectedValueException::class);
+        $this->expectExceptionMessage(
```

Comment:
I´m not a big fan of adding exception message expectations because you are coupling the test to messages than can change over time. I suggest to rely only on exceptions. If an exception changes then test has to be changed because the behavior has changed, while messages can change its meaning or be adjusted but that does not affect the behavior.

Reply:
Problem of generic exceptions like `UnexpectedValueException` is that they can be thrown multiple times during a function call and then when a test fails you don't know which exception was thrown. 
That's the reason why I add the message. I know isn't the best approach but if you have other valid ideas I'm happy to change it.